### PR TITLE
[TRTLLM-10407][perf] Enable CuteDSL indexer_top_k in model

### PR DESCRIPTION
**Pull request by limin2021**
## Description

1. add single pass multi_cta top_k kernel.
2. add config to enable cute dsl top_k in model.

### 1. perf for newly added single_pass_multi_cta kernel.

Note: 
dist: the newly added single-pass multi-cta kernel.
mc_dyn: two-pass multi-cta kernel with dynamic workload output
mc_sta: two-pass multi-cta kernel with fix-length output for 1st kernel
fi: flashinfer multi-cta kernel

dtype=float32  batch=1  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
-----------------------------------------------------------------------------------------------
        8192            7.95           12.27           11.82           11.80      1.48x      0.96x      1.00x
       16384           12.37           15.58           15.08           16.37      1.32x      1.05x      1.09x
       32768           19.88           20.73           20.36           25.09      1.26x      1.21x      1.23x
       65536           19.73           22.53           22.04           30.91      1.57x      1.37x      1.40x
      131072           19.63           26.76           26.26           34.84      1.77x      1.30x      1.33x
      262144           19.64           35.97           35.40           37.90      1.93x      1.05x      1.07x

  dtype=float32  batch=4  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            8.00           12.38           11.93           11.83      1.48x      0.96x      0.99x
       16384           12.44           15.73           15.31           16.40      1.32x      1.04x      1.07x
       32768           20.05           21.35           21.04           25.11      1.25x      1.18x      1.19x
       65536           19.94           23.36           22.83           33.37      1.67x      1.43x      1.46x
      131072           19.95           27.89           27.34           37.33      1.87x      1.34x      1.37x
      262144           20.00           37.02           36.55           40.07      2.00x      1.08x      1.10x

  dtype=float32  batch=8  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            7.98           12.40           11.98           11.83      1.48x      0.95x      0.99x
       16384           12.44           16.35           15.99           16.41      1.32x      1.00x      1.03x
       32768           20.28           21.38           21.06           25.14      1.24x      1.18x      1.19x
       65536           20.47           23.25           22.72           33.57      1.64x      1.44x      1.48x
      131072           20.25           28.15           27.57           37.45      1.85x      1.33x      1.36x
      262144           23.36           37.19           36.75           40.45      1.73x      1.09x      1.10x

  dtype=float32  batch=16  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            8.19           12.76           12.37           12.06      1.47x      0.95x      0.98x
       16384           12.69           16.68           16.30           16.67      1.31x      1.00x      1.02x
       32768           20.64           22.40           21.96           25.35      1.23x      1.13x      1.15x
       65536           20.79           23.89           23.35           33.75      1.62x      1.41x      1.45x
      131072           23.90           28.41           27.88           37.58      1.57x      1.32x      1.35x
      262144           30.23           49.08           48.47           40.75      1.35x      0.83x      0.84x

  dtype=float32  batch=32  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            8.23           12.88           12.47           12.09      1.47x      0.94x      0.97x
       16384           12.72           16.75           16.29           16.72      1.31x      1.00x      1.03x
       32768           20.98           22.82           22.36           25.48      1.21x      1.12x      1.14x
       65536           24.49           24.76           24.22           33.84      1.38x      1.37x      1.40x
      131072           30.99           40.25           39.54           37.72      1.22x      0.94x      0.95x
      262144           58.05           72.71           71.60           78.48      1.35x      1.08x      1.10x

  dtype=float32  batch=64  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            8.28           12.99           12.54           12.15      1.47x      0.94x      0.97x
       16384           12.78           17.12           16.72           16.79      1.31x      0.98x      1.00x
       32768           20.18           23.11           22.61           25.58      1.27x      1.11x      1.13x
       65536           31.96           33.09           32.63           34.19      1.07x      1.03x      1.05x
      131072           59.42           63.59           62.46           72.53      1.22x      1.14x      1.16x
      262144          113.94          110.46          108.60          116.13      1.02x      1.05x      1.07x

  dtype=float32  batch=128  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192            8.42           13.87           13.43           12.29      1.46x      0.89x      0.92x
       16384           12.91           17.35           16.94           16.99      1.32x      0.98x      1.00x
       32768           20.37           31.52           30.96           25.80      1.27x      0.82x      0.83x
       65536           61.64           51.37           50.81           65.66      1.07x      1.28x      1.29x
      131072          103.79          101.03           99.19          108.08      1.04x      1.07x      1.09x
      262144          195.64          203.19          199.99          208.18      1.06x      1.02x      1.04x

  dtype=float32  batch=256  next_n=1  top_k=2048:

  num_tokens        dist(us)      mc_dyn(us)      mc_sta(us)          fi(us)   fi/dist    fi/mcd    fi/mcs
------------------------------------------------------------------------------------------------------------------------------------
        8192           15.29           18.21           17.80           22.75      1.49x      1.25x      1.28x
       16384           24.20           25.84           25.42           31.95      1.32x      1.24x      1.26x
       32768           38.95           50.93           50.43           49.42      1.27x      0.97x      0.98x
       65536          119.26           79.23           78.63          127.70      1.07x      1.61x      1.62x
      131072          210.83          148.73          148.38          225.89      1.07x      1.52x      1.52x
      262144          350.05          286.21          285.58          372.78      1.06x      1.30x      1.31x

### 2. cute dsl top_k model accuracy check:

on mmlu/gsm8k dataset, pass the trtllm-eval check.

`[03/16/2026-13:19:43] [TRT-LLM] [I] MMLU weighted average accuracy: 87.67 (14042)`

```
[03/16/2026-13:33:06] [TRT-LLM] [I] lm-eval gsm8k results (scores normalized to range 0~100):
|Tasks|Version|     Filter     |n-shot|  Metric   |   | Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|------:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |95.1478|±  |0.5918|
|     |       |strict-match    |     5|exact_match|↑  |95.1478|±  |0.5918|

[03/16/2026-13:33:06] [TRT-LLM] [I] lm-eval gsm8k average accuracy: 95.15
```

### 3. model perf with cute dsl top_k

```
max_batch_size=32
max_num_tokens=8192            # chunked prefill: each chunk <= this size
# max_seq_len=32768              # 130K context window (KV cache capacity)
isl=32768
osl=1024
max_seq_len=isl+osl
max_num_requests=256
TP=8
EP=8

# python benchmarks/cpp/prepare_dataset.py \
#     --tokenizer=$model_card \
#     --stdout token-norm-dist \
#     --num-requests=$max_num_requests \
#     --input-mean=$isl \
#     --output-mean=$osl \
#     --input-stdev=0 \
#     --output-stdev=0 > ${DATASET}
```

image

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Review by Superjomn — APPROVED** [review]
LGTM on the llm api

File: tests/unittest/_torch/thop/parallel/test_indexer_topk.py [hunk 1]
```diff
@@ -367,17 +376,59 @@ def test_cute_dsl_topk_decode_multi_cta(
...
             top_k=index_topk,
             next_n=next_n,
+            return_val=False,
             num_copy_bits=256,
-        ),
+        )[0],
     )
+
+
+test_cute_dsl_topk_decode_distributed(
```
Comment (hyukn) [inline tests/unittest/_torch/thop/parallel/test_indexer_topk.py]: Let us remove these debugging lines before merging.

File: tests/unittest/_torch/thop/parallel/test_indexer_topk.py [hunk 2]
```diff
@@ -367,17 +376,59 @@ def test_cute_dsl_topk_decode_multi_cta(
 @pytest.mark.parametrize("index_topk", [2048])
 @pytest.mark.parametrize("num_tokens", [4096, 8192, 65536, 131072])
 def test_cute_dsl_indexer_topk_decode(batch_size, next_n, index_topk, num_tokens):
```
Comment (hyukn) [inline tests/unittest/_torch/thop/parallel/test_indexer_topk.py]: Shall we test distributed=True?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 3]
```diff
@@ -1522,10 +1531,18 @@ def sparse_attn_indexer(
                 # This is because rowEnd = seq_len - next_n + offset + 1
                 gen_kv_lens_cuda = metadata.kv_lens_cuda_runtime[
                     num_contexts:num_contexts + num_generations]
-                torch.ops.trtllm.indexer_topk_decode(
-                    logits_decode, gen_kv_lens_cuda,
-                    topk_indices_buffer[num_ctx_tokens:num_ctx_tokens +
-                                        num_gen_tokens, :], next_n)
+                # TODO: Note, cute dsl top_k requires extra memory.
+                if self.use_cute_dsl_topk and num_gen_tokens <= 256:
```
[the remark this answers is not in the document]
Comment (limin2021) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: done.

**Review by hyukn — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py [hunk 4]
```diff
@@ -2824,6 +2828,30 @@ def _get_num_sms() -> int:
         torch.float32: cutlass.Float32,
     }
 
+    def _get_or_alloc_buffer(buffer_cache, key, shape, dtype, device="cuda"):
```
[the remark this answers is not in the document]
Comment (yuxianq) [reply tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py]: Will address in a follow-up PR
